source/src/sfincs_bmi.f90

```
@@ -180,83 +180,46 @@ end function get_time_units
 
 !-----------------------------------------------------------------------------------------------------!
 
-   function get_value(var_name, ptr) result(ierr) bind(C, name="get_value")
+   function get_value(var_name, dest, n) result(ierr) bind(C, name="get_value")
       !DEC$ ATTRIBUTES DLLEXPORT :: get_value
 
+      integer(kind=c_int) :: ierr, n
       character(kind=c_char), intent(in) :: var_name(*)
-      type(c_ptr), intent(inout) :: ptr
-      integer(kind=c_int) :: ierr
+      real(c_float), dimension(n) :: dest
 
       character(len=strlen(var_name)) :: f_var_name
-      real(c_float), pointer:: f_ptr(:)
-      integer :: i
+      f_var_name = char_array_to_string(var_name, strlen(var_name))
 
       ierr = ret_code%success
 
-      ! get the name
-      f_var_name = char_array_to_string(var_name, strlen(var_name))
-
       select case(f_var_name)
        case("z_xz") ! x grid cell centre
```

Segmentation faults can happen here if the fortran arrays had not been allocated when get_value is called. I do not know the actual order of initialisation of arrays in sfnics.
What can be done here is the following:
```
case("z_xz") ! x grid cell centre
     if(allocated(z_xz)) then
       call c_f_pointer(ptr, f_ptr, [np])
       do i = 1, np
         f_ptr(i) = z_xz(i)
       end do
     else
       ierr = ret_code%failure
       ptr = c_null_ptr
     end if
```
Is the responsibility of the caller to check the return code and proceed accordingly. In the C++ wrapper this should throw an exception.